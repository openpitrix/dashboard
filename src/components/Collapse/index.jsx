import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import _ from 'lodash';

import { Icon, Switch } from 'components/Base';

import styles from './index.scss';

export default class SingleCollapse extends Component {
  static propTypes = {
    checked: PropTypes.bool,
    disabled: PropTypes.bool,
    iconPosition: PropTypes.oneOf(['left', 'right']),
    iconType: PropTypes.oneOf(['switch', 'chevron-right', 'chevron-up']),
    onChange: PropTypes.func,
    toggleType: PropTypes.oneOf(['header', 'icon'])
  };

  static defaultProps = {
    checked: false,
    disabled: false,
    iconPosition: 'left',
    iconType: 'chevron-right',
    toggleType: 'icon',
    onChange: _.noop
  };

  state = {
    isCheck: this.props.checked
  };

  renderIcon() {
    const { iconType, toggleType } = this.props;
    const { isCheck } = this.state;

    const props = {};
    if (toggleType === 'icon') {
      props.onClick = this.toggleCheck;
      props.className = styles.cursorPointer;
    }
    if (iconType === 'switch') {
      props.onChange = this.toggleCheck;
      return <Switch checked={isCheck} {...props} />;
    }

    props.type = 'dark';
    if (isCheck) {
      props.name = 'chevron-down';
    } else {
      props.name = iconType;
    }

    return <Icon {...props} />;
  }

  toggleCheck = () => {
    const isCheck = !this.state.isCheck;
    this.setState({ isCheck }, this.props.onChange(isCheck));
  };

  renderHeader() {
    const { iconPosition, header, toggleType } = this.props;
    const clickable = toggleType === 'header' || this.state.isCheck;
    const onClick = clickable ? this.toggleCheck : _.noop;
    return (
      <div
        onClick={onClick}
        className={classnames(styles.headerContainer, {
          [styles.cursorPointer]: clickable
        })}
      >
        {iconPosition === 'left' && this.renderIcon()}
        <span className={styles.header}>{header}</span>
        {iconPosition === 'right' && this.renderIcon()}
      </div>
    );
  }

  render() {
    const {
      children, className, checkCls, disabled
    } = this.props;
    const { isCheck } = this.state;
    return (
      <div
        className={classnames(
          styles.container,
          {
            [styles.disabled]: disabled,
            [styles.isCheck]: isCheck,
            [checkCls]: isCheck
          },
          className
        )}
      >
        {this.renderHeader()}
        <div className={styles.content}>{children}</div>
      </div>
    );
  }
}
