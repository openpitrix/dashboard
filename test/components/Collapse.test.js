import Collapse from 'components/Collapse';

describe('Component Collapse ', () => {
  it('basic render', () => {
    const wrapper = render(<Collapse toggleType={'header'} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
