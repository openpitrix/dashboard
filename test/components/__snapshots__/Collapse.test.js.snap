// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Collapse  basic render 1`] = `
<div
  class="container"
>
  <div
    class="headerContainer cursorPointer"
  >
    <span
      class="icon icon-chevron-right"
      style="width:16px;height:16px"
    >
      <svg
        class="qicon qicon-chevron-right qicon-dark"
      >
        <use
          href="#qui-chevron-right"
        />
      </svg>
    </span>
    <span
      class="header"
    />
  </div>
  <div
    class="content"
  />
</div>
`;
